Add new ocioview app (alpha)

michdolan opened the pull request:
`ocioview` is a visual editor for OCIO configs, written in Python. The app currently consists of two main components; a viewer and a config editor. The viewer is an expanded version of `pyociodisplay` with pan/zoom and pixel sampling support. Multiple viewers can be loaded in different tabs. The config editor is a tabbed model/view interface for the current config. Models for each config item type interface directly with the config in memory. The app's scene file is a config. This design allows dynamic interconnectivity between config items, reducing risk of errors during config authoring. Undo/redo stack support for most features is implemented.

These two components are linked with 10 possible transform subscriptions. Each subscription tracks the transform(s) for one config item, and each viewer can subscribe to any of these transforms. This provides fast visual feedback for transform editing. There is also a code view linked to the active viewer, which displays a CTF and GLSL representation of the viewer's processor, as well as the current config YAML.

This PR accounts for the core functionality needed for authoring a config, but there is still much more that could be done. I will continue working on `ocioview`, but want to invite others to contribute as well. There is plenty to refine and add, and this project is a great opportunity to improve the accessibility and interactivity of OCIO config authoring.

In terms of reviewing the code; it's a lot. There's ~15k lines of code here. There are bugs that still need to be squashed, which I am continuing to work on, but the best way to review is probably to try it out. Once the initial version is merged, we can iterate on smaller subsets of the application. Thanks for taking a look!

Some caveats:

- I developed this app on Windows, and have not done any testing on Linux or Mac. The code is cross-platform by nature of being Python, but there may be subtle issues I haven't anticipated. 
- `ocioview` depends on the OIIO and imath Python bindings, which are not currently available on PyPI. I intend to contribute a Python script to make building the dependencies simpler, but for now users will need to provide their own builds. 
- The code currently references PySide2 directly, rather than using a shim like Qt.py. I plan to update the code to use PySide6 soon, to be compatible with VFX reference platform 2024. It may make sense to adopt a shim (Qt.py or QtPy) at a later date.
- There is currently no test suite for the application itself, but this is certainly an area for improvement.
- There are some OCIO features which have not yet been implemented: grading transforms, virtual views, active/inactive state of color spaces.

Screenshots:

<img width="956" alt="ocioview_001" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/b611aa9c-1cab-490b-9115-310294ef1841">
<img width="959" alt="ocioview_015" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/4ae20f33-395c-453b-a0e7-05186d433cad">
<img width="953" alt="ocioview_018" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/03731f68-9b1b-494d-8c61-ffb31f910cc7">
<img width="960" alt="ocioview_019" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/d795b95b-b5ac-4112-83a2-6e7788a148c9">
<img width="960" alt="ocioview_020" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/455e828b-51ec-4771-a875-172b1bedc0fb">
<img width="958" alt="ocioview_021" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/30991b20-2dd7-40e5-bb12-5d8b93e9d9b2">

KelSolaar commented:
Oh exciting!

remia commented:
Looks amazing @michdolan! I haven't looked at the code in details yet but had some issues running `ocioview` on mac OS so I did some quick changes to make it work (moved to PySide6 as I had issues with PySide2 on Python 3.11, removed Imath for install simplicity, and made OIIO optional for the same reason) but no guarantee I didn't break stuff in the process. In case that can help others testing it, here is: https://github.com/remia/OpenColorIO/tree/ocioview-mac.

michdolan commented:
I made some performance improvements. Previously code for all the code views was updating with any change. Now code is only updated with code views being actively viewed, and when toggling between code views, an update will be requested immediately and then the view will receive updates on relevant changes. All code updates are asynchronous, so this lessens the amount of background processing (pygments syntax highlighting and HTML formatting) fighting the GUI thread for the GIL. 

CTF formatting is now also clearing LUT data from 1D/3D LUT transforms (which were a big bottleneck when present). Instead there is now a note where the LUT data would be, informing the user to export the CTF for the complete data. A "Export CTF" button at the top of the view can facilitate this.

<img width="417" alt="image" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/fb274b2b-b06c-4f91-96ed-c1e59b378aa8">

The shader view now also provides a box to choose the shading language to display. The viewer's language doesn't change; just the displayed code.

<img width="443" alt="image" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/7e01423b-6bb3-4415-86de-9e1cd4018fda">

michdolan commented:
Made another push that adds persistent settings for tracking recent config and image locations, along with recent file menus. Transform subscriptions now have a suffix which indicates the item type, for clarity. I also updated the interface for working with transform directionality in the viewer, also to be more clear.

I'm done making changes to this PR outside of addressing feedback. I've squashed most of the bugs I could find through manual testing. Additional work can happen in other PRs.

remia commented:
@michdolan Did you think about how you want `ocioview` to be versioned and distributed?

doug-walker reviewed (approved):
I have not had time to review everything, but I haven't found anything to comment on in that regard (your code is always excellent Michael!).  It seems like we should merge what you have here as a basis for further development.

The one thing that would be great to add would be some kind of readme to explain the current development status, what platforms it's been tested on, and what are the planned next steps.  (You have some of this in the description in the PR, but people may not see that once it's merged.)

I tried to test this on macOS using Remi's branch.  I just created a virtual env and then ran pip3 install -r requirements.txt.  The UI launched and I was able to load a config and an image but some stuff seems broken.  For example, the second of the drop-down menus seems locked to pass-through and I could not get it to load a display/view transform.  Not sure if this is a problem with Remi's branch or something I'm doing wrong (probably the latter).  We'll need to add some documentation for people.

How would you like to proceed with merging this @michdolan ?

michdolan commented:
I can address the readme/docs and versioning in a near-future PR. Once this is merged and @remia 's changes are merged, I have another PR in the works that improves a number of areas an adds a curve viewer. I can include the docs with that PR, if that's ok with you all.

remia reviewed (approved):
I haven't reviewed the PR in details but I think it's fine to merge (code quality looks really good) and improve incrementally (as needed) as Michael suggests.

michdolan commented:
README is now added. I also pushed my updates and current work on the curve viewer, which has linear and logarithmic scales. I'll have further work on it in a future PR:
<img width="333" alt="image" src="https://github.com/AcademySoftwareFoundation/OpenColorIO/assets/2299963/5d89383d-2ef1-4437-856e-9a1ae419daf6">

doug-walker commented on src/apps/ocioview/ocioview/README.md:
```diff
@@ -0,0 +1,48 @@
+* PyOpenColorIO
+* [OpenImageIO (Python bindings)](https://github.com/OpenImageIO/oiio)
+* [Imath (Python bindings)](https://github.com/AcademySoftwareFoundation/Imath)
+* ``pip install requirements.txt``
```
Suggest:
pip install -r requirements.txt

doug-walker commented on src/apps/ocioview/ocioview/README.md:
```diff
@@ -0,0 +1,48 @@
+
+``ocioview`` being an alpha release means this app is functional, but still in development, so 
+may have some rough edges. Feedback and bug reports are appreciated.
+
```
It doesn't work on Mac w/out Remi's fixes, right? Perhaps add:
Development has mostly been done on Windows. Better support for other platforms is forthcoming.
